documentation/adr/0003-use-mqtt-to-publish-trip-modifications.md

```diff
@@ -0,0 +1,26 @@
+# 3. Use MQTT to publish Trip Modifications
+
+Date: 2024-07-04
+
+## Status
+
+Proposed
+
+## Context
+
+We need to send [GTFS Trip Modifications](https://gtfs.org/realtime/feed-entities/trip-modifications/)
+to Transit Data.
+
+## Decision
+
+To minimize the amount of resources used and latency between Skate publishing or
+un-publishing a Trip Modification, we are going to use MQTT to form a 
+Event Stream between Skate and Transit Data.
+
+Since this is an output of Skate, we will not make this service required to do
+local development.
+
+## Consequences
+
+We'll be adding a new technology and service to Skate, this is more surface area
+to test.
```

Comment:
Are there any additional risks that we need to consider?

Here are a few possible ones:
1. We need to be super careful in order to ensure that Transit Data gets up-to-date detour info. If the wrong message winds up being retained, then Transit Data won't have any way to get the actual latest message.
   - If we run purely on fire-and-forget, then if a message doesn't make it to the MQTT broker for whatever reason, then it never will. 
   - If we do retry sending messages on failure, then we could wind up sending messages out of order.
1. We may need to keep stale topics around for a long time so that TD knows that a detour has been deactivated.

Reply:
> If the wrong message winds up being retained

I don't envision this as being an issue, can you elaborate on why it would be?

Also, I don't believe this risk would belong in this ADR either.

Reply:
> We may need to keep stale topics around for a long time so that TD knows that a detour has been deactivated.

I think this is also not a concern for this ADR. But I also don't think we need to keep stale topics around, we can discuss further in #2686 if you'd like.

Reply:
> I don't envision this as being an issue, can you elaborate on why it would be?

If we're not careful (and implement retry in the wrong way), we could wind up with a sequence of events like:
- Create Detour A (communicated with `message 1`)
- Sending `message 1` fails - queue a retry
- Update Detour A in some way (communicated with `message 2`)
- Sending `message 2` succeeds
- Sending `message 1` retry succeeds

Now the latest message about Detour A is `message 1`, which doesn't have whatever the updated info from `message 2` is.

Making sure that the right message is retained is a solvable problem, but any time messages are a thing, there is that bit of risk that messages will get delivered out of order.

Reply:
> If we're not careful (and implement retry in the wrong way),

What do you think about this?
```diff
diff --git a/documentation/adr/0003-use-mqtt-to-publish-trip-modifications.md b/documentation/adr/0003-use-mqtt-to-publish-trip-modifications.md
index f72ba5caf2...8cb8b8c8bc 100644
--- a/documentation/adr/0003-use-mqtt-to-publish-trip-modifications.md
+++ b/documentation/adr/0003-use-mqtt-to-publish-trip-modifications.md
@@ -57,6 +57,10 @@
 Since we'll be adding a new technology and service to Skate, there will be more
 surface area to test.
 
+By using a message queue, we risk sending events out of order and
+having inaccurate ordering of data, we'll need to be careful to address these
+issues as they arise during implementation.
+
 [^ps-1]: In this comment [Paul Swartz mentioned that using MQTT but sending all
   messages using `retain` will effectively be the same as uploading to S3, but
   with lower latency to clients](https://app.asana.com/0/0/1207124404105396/1207386033651835/f).
```